src/matrixvariates.jl

```diff
@@ -197,14 +197,12 @@ function pdf!(r::AbstractArray, d::MatrixDistribution, X::AbstractArray{M}) wher
     _pdf!(r, d, X)
 end
 
-function logpdf(d::MatrixDistribution, X::AbstractArray{M}) where M<:Matrix
-    T = promote_type(partype(d), eltype(M))
-    _logpdf!(Array{T}(undef, size(X)), d, X)
+function logpdf(d::MatrixDistribution, X::AbstractArray{<:AbstractMatrix{<:Real}})
```

Comment:
It seems to me that we shouldn't have such an implicitly broadcasting method at all. We generally deprecated these a long time ago. Is there a reason we kept this one?

Reply:
I don't know why it exists but it is definitely inconsistent that currently the application of `(log)pdf(!)` to multiple samples is deprecated for univariate distributions but not for multi- and matrixvariate distributions. In general, I think it would be good to be consistent and either allow this application for all distributions or remove all implementations. I lean towards the former, mainly because it is difficult to write generic code that computes the logpdf for multiple samples for arbitrary distributions without these `logpdf` definitions since one cannot use broadcasting to evaluate the `logpdf` for a matrix of multivariate samples in the same way.

Reply:
Couldn't you just write
```julia
julia> logpdf.(Ref(InverseWishart(2, Matrix{Float64}(I, 2, 2))), fill(Matrix{Float64}(I, 2, 2), 2))
2-element Array{Float64,1}:
 -3.5310242469692907
 -3.5310242469692907
```

Reply:
Sure, you can do this in this case. And you can do it in the same way for univariate distributions:
```julia
julia> logpdf.(Ref(Normal()), rand(Normal(), 2))
2-element Array{Float64,1}:
 -2.413956289132882
 -1.0116727704190984
```
Unsurprisingly though, the following fails:
```julia
julia> logpdf.(Ref(MvNormal(2, 1)), rand(MvNormal(2, 1), 2))
ERROR: MethodError: no method matching logpdf(::MvNormal{Float64,PDMats.ScalMat{Float64},FillArrays.Zeros{Float64,1,Tuple{Base.OneTo{Int64}}}}, ::Float64)
...
```
So it is not possible to generically use `logpdf.(dist, samples)` to compute logpdfs of samples but instead one has to dispatch on the type of distributions and samples and either add a custom implementation for these cases or call
```julia
julia> logpdf(MvNormal(2, 1), rand(MvNormal(2, 1), 2))
2-element Array{Float64,1}:
 -2.0814343848408745
 -3.5913383903349345
```

As I mentioned above, I would be fine with either removing all of these `logpdf` implementations for multiple samples or adding them for all distributions, as long as it is done in a consistent way. However, I think it would be easier if Distributions would implement `logpdf` for multiple samples for all distributions since otherwise every downstream package that wants to handle logpdf calculations for all distribution types has to add these implementations and make sure that broadcasting is used for univariate and matrixvariate but not for multivariate distributions + matrices of samples.

Reply:
I really think we should avoid implicit broadcasting. We've spent quite some time teaching people to be explicit about broadcasting and it's also the only way to hope for composability of broadcasting operations. IMO, the issue here is the return type of the multivariate distributions which doesn't fit well into the general interface. Until we have a general solution for that, I think the better approach is to ask people to use `eachrow` explicitly for the multivariates, i.e. something like
```julia
julia> logpdf.(Ref(MvNormal(ones(2))), eachrow(randn(10,2)))
10-element Array{Float64,1}:
 -3.2672567129269323
 -2.6944444370392286
 -4.083857499332683
 -2.221300852511519
 -1.9903141948994234
 -3.009577938574812
 -4.12761318469939
 -3.0240385144021076
 -2.0975290149847705
 -2.5841749740831976
 ```

Reply:
OK, so should all these `(log)pdf(!)` be deprecated in this PR? Or better in a separate PR? I think regardless of these deprecations the in-place calls are not "correct" and should be avoided.

Reply:
I'm fine with either doing it as part of this PR or in a follow-up you believe it simplifies things.